tests/e2e/src/main.rs

```diff
@@ -0,0 +1,219 @@
+    let connection = create_connection(&config).set_signer(sudo);
+
+    let validators_before: Vec<AccountId> = connection
+        .get_storage_value("Session", "Validators", None)?
+        .unwrap();
+
+    info!("[+] Validators before tx: {:#?}", validators_before);
+
+    let mut new_validators: Vec<AccountId> = config
+        .accounts
+        .into_iter()
+        .map(|pair| pair.public().into())
+        .collect();
+
+    new_validators.push(
+        AccountId::from_ss58check("5EHkv1FCd4jeQmVrbYhrETL1EAr8NJxNbukDRT4FaYWbjW8f").unwrap(),
+    );
+
+    info!("[+] New validators {:#?}", new_validators);
+
+    // what session is this?
+
+    let session_number: u32 = connection
+        .get_storage_value("Session", "CurrentIndex", None)
+        .unwrap()
+        .unwrap();
+
+    info!("[+] Session number {:?}", session_number);
+
+    let call = compose_call!(
+        connection.metadata,
+        "Aleph",
+        "change_validators",
+        new_validators.clone(),
+        session_number + 1
+    );
+
+    let tx = compose_extrinsic!(connection, "Sudo", "sudo_unchecked_weight", call, 0_u64);
+
+    // send and watch extrinsic until finalized
+    let tx_hash = connection
+        .send_extrinsic(tx.hex_encode(), XtStatus::InBlock)
+        .expect("Could not send extrinsc")
+        .expect("Could not get tx hash");
+
+    info!("[+] Transaction hash: {}", tx_hash);
+
+    // wait for the event
+
+    let block_hash = connection.get_finalized_head()?.unwrap();
+    let header: Header = connection.get_header(Some(block_hash))?.unwrap();
+    let block_number = header.number;
+
+    info!("[+] Current block number: {}", block_number);
+
+    // NOTE: this is hackish, we are assuming that two blocks is enough
+    // ideally we should wait until the event arrives
+    // see `api::subscribe_events`
+    // but I could not readily get it to work with this event
+    let _current_block_number = wait_for_block(connection.clone(), header.number + 2)?;
+
+    let validators_after: Vec<AccountId> = connection
+        .get_storage_value("Session", "Validators", None)?
+        .unwrap();
+
+    info!("[+] Validators after tx: {:#?}", validators_after);
+
+    assert!(are_equal(new_validators.clone(), validators_after));
+
+    Ok(())
+}
+
+/// blocks the main thread waiting for a block with a number at least `block_number`
+fn wait_for_block(
+    connection: Api<sr25519::Pair, WsRpcClient>,
+    block_number: u32,
+) -> anyhow::Result<u32> {
+    let (sender, receiver) = channel();
+    connection.subscribe_finalized_heads(sender)?;
+
+    while let Ok(header) = receiver
+        .recv()
+        .map(|h| serde_json::from_str::<Header>(&h).unwrap())
+    {
+        info!("[+] Received header for a block number {:?}", header.number);
+
+        if header.number.ge(&block_number) {
+            return Ok(block_number);
+        }
+    }
+
+    Err(anyhow::anyhow!("Giving up"))
+}
+
+fn are_equal<T>(a: Vec<T>, b: Vec<T>) -> bool
```

Comment:
The order should be the same, right? so `&a == &b` should be enough.

Reply:
removed